sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -1,24 +1,56 @@
 // Copyright 2021-2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use alloc::{boxed::Box, string::String, vec::Vec};
-use core::{ops::RangeInclusive, str::FromStr};
-
-use packable::{bounded::BoundedU16, prefix::BoxedSlicePrefix};
+use alloc::{
+    collections::BTreeMap,
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::ops::{Deref, RangeInclusive};
+
+use packable::{
+    bounded::{BoundedU16, BoundedU8},
+    prefix::{BTreeMapPrefix, BoxedSlicePrefix},
+};
 
 use crate::types::block::{output::StorageScore, protocol::WorkScore, Error};
 
 pub(crate) type MetadataFeatureLength =
     BoundedU16<{ *MetadataFeature::LENGTH_RANGE.start() }, { *MetadataFeature::LENGTH_RANGE.end() }>;
 
+pub(crate) type MetadataFeatureKeyLength = BoundedU8<1, { u8::MAX }>;
+pub(crate) type MetadataFeatureValueLength = BoundedU16<0, { u16::MAX }>;
+
 /// Defines metadata, arbitrary binary data, that will be stored in the output.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, packable::Packable)]
-#[packable(unpack_error = Error, with = |err| Error::InvalidMetadataFeatureLength(err.into_prefix_err().into()))]
+#[packable(unpack_error = Error, with = |err| Error::InvalidMetadataFeature(err.to_string()))]
 pub struct MetadataFeature(
     // Binary data.
-    pub(crate) BoxedSlicePrefix<u8, MetadataFeatureLength>,
+    #[packable(verify_with = verify_keys_packable)]
+    pub(crate)  BTreeMapPrefix<
+        BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+        BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        MetadataFeatureLength,
+    >,
 );
 
+fn verify_keys_packable<const VERIFY: bool>(
+    map: &BTreeMapPrefix<
+        BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+        BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        MetadataFeatureLength,
+    >,
+) -> Result<(), Error> {
+    if VERIFY {
+        for key in map.keys() {
```

Comment:
check (according to TIP) for at least one entry in the map not necessary?

Reply:
Doesn't `LENGTH_RANGE` already ensure that during unpacking? Or do you see a way how it can pass that without a key?

Reply:
It could pass it with just the key length prefix (but not the actual key)?

Reply:
Don't think so, since the key also is also bounded `MetadataFeatureKeyLength = BoundedU8<1, { u8::MAX }>` and then during unpacking it would fail since it's missing >=1 byte

Reply:
hmm, I guess in combination of both you're right and this is enough to ensure there's at least 1 entry.

Reply:
Added